compiler/modules/front.syntax/tests/posTree.bal

```diff
@@ -0,0 +1,585 @@
+    RootSyntaxNode root = buildTree(part);
+    foreach SyntaxNode node in root.childNodes {
+        validateSyntaxNode(node);
+    }
+}
+
+function validateSyntaxNode(SyntaxNode node) {
+    if node is FixedSyntaxNode|IdentifierSyntaxNode {
+        return;
+    }
+    else {
+        Position parentStart = node.astNode.startPos;
+        Position parentEnd = node.astNode.endPos;
+        if node is NonTerminalSyntaxNode {
+            Position lastEnd = parentStart;
+            foreach SyntaxNode child in node.childNodes {
+                if child is TerminalSyntaxNode|NonTerminalSyntaxNode {
+                    Position childStartPos = child.astNode.startPos;
+                    Position childEndPos = child.astNode.endPos;
+                    if childStartPos < parentStart || childEndPos > parentEnd {
+                        outofBoundChildErrr(node, child);
+                    }
+                    if childStartPos < lastEnd {
+                        overlappingNodeErr(node, lastEnd, childStartPos);
+                    }
+                    // pr-to: check if difference is all white space
+                    lastEnd = childEndPos;
+                    validateSyntaxNode(child);
+                }
+                else {
+                    Position? pos = child.pos;
+                    if pos != () {
+                        if pos < lastEnd {
+                            overlappingNodeErr(node, lastEnd, pos);
+                        }
+                        lastEnd = pos;
+                    }
+                    // pr-todo: consume token
+                }
+            }
+        }
+    }
+}
+
+// pr-todo: show file names in the error msg
+function overlappingNodeErr(NonTerminalSyntaxNode parent, Position lastEnd, Position currentStart) {
+    string[] body = ["overlapping child nodes"];
+    body.push(string `ast node: ${parent.astNode.toString()}`);
+    body.push(string`overlapping range: ${unpackPosition(lastEnd).toString()}:${unpackPosition(currentStart).toString()}`);
+    string msg = "\n".'join(...body);
+    panic error(msg);
+}
+
+function outofBoundChildErrr(NonTerminalSyntaxNode parent, NonTerminalSyntaxNode|TerminalSyntaxNode child) {
+    string[] body = ["child node outside of parent"];
+    body.push(string `parent node: ${parent.astNode.toString()}`);
+    body.push(string `child node: ${child.astNode.toString()}`);
+    string msg = "\n".'join(...body);
+    panic error(msg);
+}
+
+function buildTree(ModulePart part) returns RootSyntaxNode {
+    SyntaxNode[] importDecls = from ImportDecl decl in part.importDecls select importDeclToNode(decl);
+    SyntaxNode[] moduleLevelDefns = from ModuleLevelDefn defn in part.defns select moduleLevelDefnToNode(defn);
+    SyntaxNode[] childNodes = [];
+    childNodes.push(...importDecls);
+    childNodes.push(...moduleLevelDefns);
+    return { childNodes, part };
+}
+
+function importDeclToNode(ImportDecl decl) returns NonTerminalSyntaxNode {
```

Comment:
Use the pattern `syntaxNodeFromXYZ` (so everything starts with `syntaxNode`), and put the syntax node tree construction in a separate file `syntaxNode`.

Reply:
fixed